src/Analysis/Ast/Impl/Extensions/NodeExtensions.cs

```diff
@@ -46,5 +49,72 @@ public static class NodeExtensions {
             }
             return e;
         }
+
+        public static List<Node> GetAncestorsOrThis(this Node root, Node node, CancellationToken cancellationToken) {
+            var parentChain = new List<Node>();
+
+            // there seems no way to go up the parent chain. always has to go down from the top
+            while (root != null) {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var temp = root;
+                root = null;
+
+                // this assumes node is not overlapped and children are ordered from left to right
+                // in textual position
+                foreach (var current in GetChildNodes(temp)) {
```

Comment:
`GetChildNodes(temp).Where(x => x.IndexSpan.Contains(node.IndexSpan))` but up to you